src/compute/src/sink/materialized_view_v2.rs

```diff
@@ -33,5 +155,875 @@ pub(super) fn persist_sink<S>(
+                self.batches_frontier = frontier;
+            }
+        }
+
+        /// Advance the current `lower`.
+        ///
+        /// Discards all currently stashed batches and batch descriptions, assuming that they are
+        /// now invalid.
+        async fn advance_lower(&mut self, frontier: Antichain<Timestamp>) {
+            debug_assert!(PartialOrder::less_than(&self.lower, &frontier));
+
+            self.lower = frontier;
+            self.batch_description = None;
+
+            // Remove stashed batches, cleaning up those we didn't append.
+            for batch in self.batches.drain(..) {
+                batch.delete().await;
+            }
+        }
+
+        /// Absorb the given batch description into the state, provided it is not outdated.
+        async fn absorb_batch_description(&mut self, desc: BatchDescription) {
+            if PartialOrder::less_than(&self.lower, &desc.lower) {
+                self.advance_lower(desc.lower.clone()).await;
+            } else if &self.lower != &desc.lower {
+                return;
+            }
+
+            self.batch_description = Some(desc);
+        }
+
+        /// Absorb the given batch into the state, provided it is not outdated.
+        async fn absorb_batch(&mut self, batch: ProtoBatch) {
+            let batch = self.persist_writer.batch_from_transmittable_batch(batch);
+            if PartialOrder::less_than(&self.lower, batch.lower()) {
+                self.advance_lower(batch.lower().clone()).await;
+            } else if &self.lower != batch.lower() {
+                return;
+            }
+
+            self.batches.push(batch);
+        }
+
+        async fn maybe_append_batches(&mut self) {
+            let batches_complete = PartialOrder::less_than(&self.lower, &self.batches_frontier);
+            if !batches_complete {
+                return;
+            }
+
+            let Some(desc) = self.batch_description.take() else {
+                return;
+            };
+
+            let new_lower = match self.append_batches(desc).await {
+                Ok(shard_upper) => shard_upper,
+                Err(shard_upper) => {
+                    // Failing the append is expected in the presence of concurrent replicas. There
+                    // is nothing special to do here: The self-correcting feedback mechanism
+                    // ensures that we observe the concurrent changes, compute their consequences,
+                    // and append them at a future time.
+                    shard_upper
+                }
+            };
+
+            self.advance_lower(new_lower).await;
+        }
+
+        /// Append the current `batches` to the output shard.
+        ///
+        /// Returns whether the append was successful or not, and the current shard upper in either
+        /// case.
+        ///
+        /// This method advances the shard upper to the batch `lower` if necessary. This is the
```

Comment:
Should this only happen when the shard upper is T::minimum? If so, we could model it like we do in CTs where it's done, if necessary, at operator startup

Reply:
In an ordered world we could add an assert here for `T::minimum()` yes. In reality it happens sometimes that we restart an MV dataflow with an `as_of` that's greater than the current `upper` of its output shard. I don't remember the details anymore, but there was an incident where a persist shard was corrupted and we had to manually drop it. This allowed the environment to restart and recreate the shard, but the new shard didn't contain the historical detail required by an MV depending on it, so the new MV `as_of` would be later than its current `upper`. You could argue that we should always crash to avoid correctness issues, but sometimes users care more about availability, or don't care about this specific instance of incorrectness at all.

Apart from folk manually mucking with the shards, there are also more legitimate reasons to want to skip times in the MV output. We have discussed in the past that if we know nobody is observing the output of an MV for some time, it should be fine to forward its write frontier across that time. That might be handy to allow compaction of the dependencies of an MV that's installed on an unhealthy/unprovisioned cluster.

So my thinking is that the MV sink should always start writing to the output shard at its provided `as_of`, regardless of the current shard frontier.

But even without a `T::minimum()` check we could still do the advancing to the `as_of` at operator startup. The reason I'm not doing that is that it complicates things with read-only mode. In read-only mode we shouldn't write, so we need to defer the advancing to the point in time where read-only mode is switched off. With the current implementation it works out that only the `mint` operator checks the read-only flag and suppresses batch minting when it is set. Once read-only mode is switched off, the first batch is minted, reaches the `append` operator, which bumps the shard frontier to the `as_of` as part of trying to append the batch.

Reply:
TIL!

(This is all great context to copy in here as code comments, btw, if you feel so inclined)

Reply:
Added some more text about read-only mode to this method and the module docs.